### PR TITLE
Improve schemas

### PR DESCRIPTION
**Pull request by codeworse**
closing https://github.com/serenedb/serenedb/issues/34

### CODE REVIEW

File: server/pg/pg_comm_task.cpp [hunk 1]
```diff
@@ -312,6 +312,9 @@ void PgSQLCommTaskBase::HandleClientHello(std::string_view packet) {
         SendParameterStatus(param, GetDefaultVariable(param));
       }
 
+      _connection_ctx->GetConfig().Set(
+        Config::VariableContext::Session, "current_database_id",
+        folly::to<std::string>(_database->GetId().id()));
```
Comment (MBkkt) [inline server/pg/pg_comm_task.cpp]: is it pg specific parameter?

File: server/catalog/sql_function_impl.cpp [hunk 2]
```diff
@@ -81,10 +81,13 @@ Result FunctionImpl::Init(ObjectId database, std::string_view name,
   if (!r.ok()) {
     return r;
   }
+  auto search_path = Config()
+                       .Get<VariableType::PgSearchPath>("search_path")
+                       .value_or(std::vector<std::string>{});
```
Comment (MBkkt) [inline server/catalog/sql_function_impl.cpp]: is it possible that we have nullopt here?

File: server/catalog/sql_query_view.cpp [hunk 3]
```diff
@@ -98,10 +98,15 @@ Result SqlQueryViewImpl::Check(ObjectId database, std::string_view name,
             "sql query view should contains select statement"};
   }
 
+  auto search_path = Config()
+                       .Get<VariableType::PgSearchPath>("search_path")
+                       .value_or(std::vector<std::string>{});
```
Comment (MBkkt) [inline server/catalog/sql_query_view.cpp]: same

File: tests/sqllogic/any/pg/system/functions.test [hunk 4]
```diff
@@ -1,3 +1,7 @@
+statement ok
```
Comment (gnusi) [inline tests/sqllogic/any/pg/system/functions.test:1]: Let's add a test for the corner case:
* drop all schemas
   -> `SELECT current_schema()` returns `NULL``
   -> `CREATE TABLE t(i INT)` fails with the same error as in postgres

File: server/pg/connection_context.h [hunk 5]
```diff
@@ -37,6 +37,7 @@ class ConnectionContext : public ExecContext {
   }
 
   Config& GetConfig() { return _config; }
+  const Config& GetConfig() const { return _config; }
```
Comment (MBkkt) [inline server/pg/connection_context.h]: deducing this

File: server/pg/sql_resolver.cpp [hunk 6]
```diff
@@ -54,14 +55,28 @@ void ResolveObject(ObjectId database, Objects& objects, Disallowed& disallowed,
...
-    std::string_view schema = StaticStrings::kPublic;
+
     std::string_view key = name.relation;
-    if (auto object = catalog.GetTable(database, schema, key)) {
-      data.object = object;
-    } else if (auto object = catalog.GetView(database, schema, key)) {
-      data.object = object;
+
+    auto resolve_object = [&](std::string_view schema) {
+      if (auto object = catalog.GetTable(database, schema, key)) {
```
Comment (MBkkt) [inline server/pg/sql_resolver.cpp:62]:
```suggestion
    auto resolve_object = [&](std::string_view schema) {
      if (schema  == pg_catalog || schema  == information_schema) {
        return;
      } 
      if (auto object = catalog.GetTable(database, schema, key)) {
```

Or better: user cannot create pg_catalog/information_schema schemas

File: server/pg/sql_resolver.cpp [hunk 7]
```diff
@@ -54,14 +55,28 @@ void ResolveObject(ObjectId database, Objects& objects, Disallowed& disallowed,
...
+        data.object = object;
+      } else if (auto object = catalog.GetView(database, schema, key)) {
+        data.object = object;
+      } else {
+        data.object = catalog.GetFunction(database, schema, key);
+      }
+    };
+
+    if (!name.schema.empty()) {
+      resolve_object(std::string_view{name.schema});
```
Comment (MBkkt) [inline server/pg/sql_resolver.cpp]:
```suggestion
      resolve_object(name.schema);
```

File: server/pg/sql_resolver.cpp [hunk 8]
```diff
@@ -34,7 +34,8 @@ namespace sdb::pg {
 
 namespace {
 
-void ResolveObject(ObjectId database, Objects& objects, Disallowed& disallowed,
+void ResolveObject(ObjectId database, std::span<std::string> search_path,
```
Comment (MBkkt) [inline server/pg/sql_resolver.cpp]:
```suggestion
void ResolveObject(ObjectId database, std::span<const std::string> search_path,
```

File: server/pg/sql_resolver.cpp [hunk 9]
```diff
@@ -98,30 +115,34 @@ void ResolveEntity(ObjectId database, Objects& objects, Disallowed& disallowed,
...
+void ResolveQueryView(ObjectId database, std::span<std::string> search_path,
+                      Objects& objects, Disallowed& disallowed,
+                      const Objects& query) {
+  ResolveEntity(database, search_path, objects, disallowed, query, "view");
 }
 
-void ResolveFunction(ObjectId database, Objects& objects,
-                     Disallowed& disallowed, const Objects& query) {
-  ResolveEntity(database, objects, disallowed, query, "function");
+void ResolveFunction(ObjectId database, std::span<std::string> search_path,
```
Comment (MBkkt) [inline server/pg/sql_resolver.cpp]: `std::span<const std::string>`

File: server/pg/sql_resolver.cpp [hunk 10]
```diff
@@ -73,23 +88,25 @@ void ResolveObject(ObjectId database, Objects& objects, Disallowed& disallowed,
...
+      ResolveFunction(database, search_path, objects, disallowed,
                       func.SqlFunction().GetObjects());
       changed = disallowed.erase(name) != 0;
       SDB_ASSERT(changed);
     }
   }
 }
 
-void ResolveEntity(ObjectId database, Objects& objects, Disallowed& disallowed,
+void ResolveEntity(ObjectId database, std::span<std::string> search_path,
```
Comment (MBkkt) [inline server/pg/sql_resolver.cpp]:
```suggestion
void ResolveEntity(ObjectId database, std::span<const std::string> search_path,
```

File: server/query/config.h [hunk 11]
```diff
@@ -102,6 +103,8 @@ class Config : public velox::config::IConfig {
         }
         return std::optional{result};
       });
+      SDB_ASSERT(value.has_value());
```
Comment (MBkkt) [inline server/query/config.h]:
```suggestion
      SDB_ASSERT(value);
```

File: server/query/config.h [hunk 12]
```diff
@@ -102,6 +103,8 @@ class Config : public velox::config::IConfig {
         }
         return std::optional{result};
```
Comment (MBkkt) [inline server/query/config.h:105]:
```suggestion
        return result;
```
maybe?

File: server/pg/functions.cpp [hunk 13]
```diff
@@ -208,26 +210,55 @@ template<typename T>
...
+    const std::vector<velox::TypePtr>& /*inputTypes*/,
+    const velox::core::QueryConfig& config,
+    const arg_type<bool>& /*include_implicit*/) {
+    auto cfg = basics::downCast<const Config>(config.config());
+    auto database_id = cfg->GetCurrentDatabase();
+    auto search_path = cfg->Get<VariableType::PgSearchPath>("search_path");
+    auto& catalog =
+      SerenedServer::Instance().getFeature<catalog::CatalogFeature>().Global();
+    auto filter = [&](const std::string_view schema_name) {
+      return catalog.GetSchema(database_id, schema_name) != nullptr;
```
Comment (MBkkt) [inline server/pg/functions.cpp:245]:
```suggestion
      return catalog.GetSchema(database_id, schema_name);
```

File: server/pg/functions.cpp [hunk 14]
```diff
@@ -208,26 +210,55 @@ template<typename T>
...
-    // TODO(codeworse): implement proper schema resolution
-
     if (include_implicit) {
       out.add_item().copy_from("pg_catalog");
     }
-    out.add_item().copy_from(StaticStrings::kPublic);
+    absl::c_for_each(_schema_names,
+                     [&](const std::string& schema_name) mutable {
+                       out.add_item().copy_from(schema_name);
+                     });
```
Comment (MBkkt) [inline server/pg/functions.cpp]: why not `for (const auto& schema_name : schema_names)`

File: server/query/config.h [hunk 15]
```diff
@@ -159,6 +171,8 @@ class Config : public velox::config::IConfig {
   containers::FlatHashMap<std::string_view, TxnVariable> _transaction;
 
   bool _inside_transaction = false;
+  ObjectId _database_id = ObjectId::none();
+  std::string _current_schema = "";
```
Comment (MBkkt) [inline server/query/config.h]:
```suggestion
```

File: server/query/config.h [hunk 16]
```diff
@@ -137,6 +140,15 @@ class Config : public velox::config::IConfig {
 
   bool InsideTransaction() const { return _inside_transaction; }
 
+  void SetCurrentDatabase(ObjectId database_id) {
+    _database_id = database_id;
+    _current_schema = "";
+  }
+
+  ObjectId GetCurrentDatabase() const { return _database_id; }
```
Comment (MBkkt) [inline server/query/config.h]: let's get this from current execution context

File: server/pg/functions.cpp (hunk 13 above)
Comment (codeworse) [reply server/pg/functions.cpp:245]: it should be explicit or via .get()

File: server/query/config.h (hunk 12 above)
Comment (codeworse) [reply server/query/config.h:105]: it should have type `std::optional<std::vector<std::string>>`, so either specify in lambda result or mark explicitly optional

File: server/pg/sql_resolver.cpp (hunk 6 above)
Comment (codeworse) [reply server/pg/sql_resolver.cpp:62]: Shouldn't it be into `create_schema.cpp`?
